### PR TITLE
[CHORE] [V3] Refactor Initialization APIs

### PR DESCRIPTION
**Pull request by marco-saia-datadog**
### What does this PR do?

- Refactored configurations into separate modules
  - Split configuration logic into multiple files and a nested folder structure.
  - This makes it easier to navigate between different supported SDK initialization types.

- Support for initializing individual features with extra options
  - Feature configurations (core, RUM, logs, trace) now accept all required options as positional constructor arguments.
  - All optional configuration values are passed as the final positional argument as an object, allowing easier extensibility and clearer APIs.

- Added missing properties in `FileBasedConfiguration` and `AutoInstrumentationConfiguration`
  - Fixed multiple configuration properties that were not being picked up from non-standard configs.
  - Updated tests to cover all supported configuration options and ensure correct mapping.

- Implemented missing JSON converters in FileBasedConfiguration
  - Fixed cases where non-primitive values (e.g. enums) were assigned as raw values, which could lead to runtime issues.
  - All non-primitive types are now correctly parsed from the JSON configuration file.

- Added missing tests and adjusted behavior
  - Introduced additional tests to cover edge cases and updated existing ones to reflect the new configuration behavior.
  
- Removed unrelated cyclic dependency in network-instrumentation layer

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)
- [ ] If this PR is auto-generated, please make sure also to manually update the code related to the change

### CODE REVIEW

File: packages/core/src/config/features/CoreConfiguration.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+    // Logs Configuration
+    public logsConfiguration?: LogsConfiguration = DEFAULTS.logsConfiguration;
+
+    // Proxy Configuration
+    public proxyConfiguration?: ProxyConfiguration =
+        DEFAULTS.proxyConfiguration;
+
+    // RUM Configuration
+    public rumConfiguration?: RumConfiguration = DEFAULTS.rumConfiguration;
```
Comment (sbarrio) [inline packages/core/src/config/features/CoreConfiguration.ts:70]: ~~Won't we need to change this **not** to have a default version once we move forward with modularization?~~

~~If we keep it as it is, and a user just wants to initialize the Core SDK and then just the Logs module (for example), if the core config has a default RumConfiguration it would try ton initialize that too, doing something that the user didn't want to do, right?~~

Scratch that, I just saw that the default configuration values for the modules is set to undefined 😅

File: packages/core/src/config/features/CoreConfiguration.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        options: CoreConfigurationOptions = {}
+    ) {
+        const {
+            rumConfiguration,
+            logsConfiguration,
+            traceConfiguration,
+            ...rest
+        } = options;
+
+        Object.assign(this, rest);
```
Comment (sbarrio) [inline packages/core/src/config/features/CoreConfiguration.ts:107]: Much cleaner than my version 👏

File: packages/core/src/sdk/DatadogProvider/__tests__/initialization.test.tsx [hunk 3]
```diff
@@ -65,7 +64,7 @@ describe('DatadogProvider', () => {
             expect(receivedConfiguration).toMatchInlineSnapshot(`
                 DdSdkNativeConfiguration {
                   "additionalConfiguration": {
-                    "_dd.react_native_version": "${reactNativeVersion}",
+                    "_dd.react_native_version": "0.76.9",
```
Comment (sbarrio) [inline packages/core/src/sdk/DatadogProvider/__tests__/initialization.test.tsx]: Won't hardcoding this break the test?

File: packages/core/src/__tests__/DdSdkReactNativeConfiguration.test.ts [hunk 4]
```diff
@@ -156,13 +178,10 @@ describe('DdSdkReactNativeConfiguration', () => {
...
                   "clientToken": "fake-client-token",
                   "env": "fake-env",
-                  "firstPartyHosts": [
-                    "api.com",
-                  ],
                   "logsConfiguration": LogsConfiguration {
                     "bundleLogsWithRum": true,
                     "bundleLogsWithTraces": true,
-                    "customEndpoint": "https://trace.example.com/",
+                    "customEndpoint": "https://logs.example.com/",
```
Comment (sbarrio) [inline packages/core/src/__tests__/DdSdkReactNativeConfiguration.test.ts:184]: Woops 😓 

Nice catch!

File: packages/core/src/sdk/DatadogProvider/__tests__/initialization.test.tsx (hunk 3 above)
Comment (marco-saia-datadog) [reply packages/core/src/sdk/DatadogProvider/__tests__/initialization.test.tsx]: Yes, definitely! Good catch.

File: packages/core/src/config/FileBasedConfiguration.ts [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+        const jsonConfiguration = getJSONConfiguration(params?.configuration);
+
+        const {
+            clientToken,
+            env,
+            trackingConsent,
+            ...rest
+        } = jsonConfiguration;
+
+        // TODO: Handle undefined clientToken, env or trackingConsent
```
Comment (cdn34dd) [inline packages/core/src/config/FileBasedConfiguration.ts]: Should we handle this before merging ? Or is this something that can go afterwards ? If not we should at create a JIRA ticket for this and add the ticket id here.

File: packages/core/src/config/FileBasedConfiguration.ts [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+
+                this.rumConfiguration.errorEventMapper =
+                    params?.errorEventMapper ?? RUM_DEFAULTS.errorEventMapper;
+                this.rumConfiguration.resourceEventMapper =
+                    params?.resourceEventMapper ??
+                    RUM_DEFAULTS.resourceEventMapper;
+                this.rumConfiguration.actionEventMapper =
+                    params?.actionEventMapper ?? RUM_DEFAULTS.actionEventMapper;
+            } else {
+                console.warn(
```
Comment (cdn34dd) [inline packages/core/src/config/FileBasedConfiguration.ts:96]: This seems like it could go unnoticed quite easily, should we not throw an error here instead ?

File: packages/core/src/DdSdkReactNative.tsx [hunk 7]
```diff
@@ -428,21 +426,29 @@ export class DdSdkReactNative {
...
             rumConfiguration.longTaskThresholdMs = adaptLongTaskThreshold(
                 longTaskThresholdMs
             );
+
+            const nativeLongTaskThresholdMs =
+                configuration.rumConfiguration?.nativeLongTaskThresholdMs ||
+                false;
+            rumConfiguration.nativeLongTaskThresholdMs = adaptLongTaskThreshold(
+                nativeLongTaskThresholdMs
+            );
```
Comment (cdn34dd) [inline packages/core/src/DdSdkReactNative.tsx:440]: Why are we defaulting to false here ? 

```js
const nativeLongTaskThresholdMs =
                configuration.rumConfiguration?.nativeLongTaskThresholdMs ||
                false;
```
Seems this should be a number only property from the get go, because on the line after we call this function: `adaptLongTaskThreshold`, which puts the value to 0 if `nativeLongTaskThresholdMs` is falsy.

Comment (cdn34dd) [reply packages/core/src/DdSdkReactNative.tsx:440]: Also seems that `RUM_DEFAULTS` has:
```js
{
...
nativeLongTaskThresholdMs: 200,
...
}
```

So why are we defaulting to `false` which then gets translated to `0`, if the default is 200 ?

**Review by cdn34dd — DISMISSED** [review]
Overall this looks good, just left a couple of comments regarding some implementation details.

File: packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/distributedTracingAttributes.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0. This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
```
Comment (cdn34dd) [inline packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/distributedTracingAttributes.ts:5]: **nit**: keep consistency with other files previously created, this applies to other files also.

```
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */
```

File: packages/core/src/config/FileBasedConfiguration.ts (hunk 5 above)
Comment (sbarrio) [reply packages/core/src/config/FileBasedConfiguration.ts]: Yeah, we had a check inside getJsonConfiguration but I have moved all checks to this level, and modified tests accordingly.

File: packages/core/src/config/FileBasedConfiguration.ts (hunk 6 above)
Comment (sbarrio) [reply packages/core/src/config/FileBasedConfiguration.ts:96]: I think a warning should be fine and less intrusive to users integrating the SDK.  It's also the approach we follow on most places, we notify the user that the SDK is not properly configured and that it will not work in this state, but we don't block anything.

Throwing an error may impede them working properly while testing theirs apps locally and without all the configuration being available.

File: packages/core/src/DdSdkReactNative.tsx (hunk 7 above)
Comment (sbarrio) [reply packages/core/src/DdSdkReactNative.tsx:440]: Yeah, this was a leftover from the previous configuration but I agree that it does not make much sense anymore, especially when the native layer only accepts numbers and takes this being 0 as an effective way to disable longTask tracking.

I removed the boolean and adapted the code and tests to only work with numbers.

File: packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/distributedTracingAttributes.ts (hunk 8 above)
Comment (sbarrio) [reply packages/core/src/rum/instrumentation/resourceTracking/distributedTracing/distributedTracingAttributes.ts:5]: Nice catch!, it's fixed now.

**Review by cdn34dd — DISMISSED** [review]

**Review by cdn34dd — DISMISSED** [review]

**Review by cdn34dd — APPROVED** [review]
